### ISSUE TITLE
[enhancement] Block search performance: export the internal `restartInterval` configuration to the `Options` structure

### ISSUE THREAD

**Issue by schomatis** [dgraph-io/badger#446]
# Summary

The rationale behind this request is that in the IPFS project (where Badger is intended to be the default data store in the near future) the keys are hashes (of the content being stored) and hence are (ideally) uniformly distributed with maximum distance in its key space, so the objective of this issue would be to be able to decrement the `restartInterval` value (by including it in Badger's [`Options`](https://github.com/dgraph-io/badger/blob/a39f9e06de643cd1edbc85492b18fec552fc7c7e/options.go#L52)) in production systems to minimize block searches (expecting a performance improvement).

# Detailed analysis

From my understanding of the code (please correct me if necessary), the keys in the SSTables are distributed in a two layer system. In the higher layer keys are grouped in blocks with each block exposing only one key that represents the group. In the lower layer, inside each block, it can be seen that the exposed key is the initial (lower) key of the group (`baseKey`) serving as a reference key (pivot in the searches) from where the rest of the keys of the blocks are derived, encoded as differences (`keyDiff`) from the `baseKey`. The objective behind this encoding system  is to reduce storage space by compressing the keys at the cost of CPU power (see later). The number of keys in a block is set by the [`restartInterval`](https://github.com/dgraph-io/badger/blob/a39f9e06de643cd1edbc85492b18fec552fc7c7e/table/builder.go#L30) variable which is hard-coded in the logic and can't be modified by the user (is not part of the API).

When a search is performed ([`seekFrom`](https://github.com/dgraph-io/badger/blob/a39f9e06de643cd1edbc85492b18fec552fc7c7e/table/iterator.go#L275)), first a (higher layer) binary search ([`sort.Search`](https://github.com/dgraph-io/badger/blob/a39f9e06de643cd1edbc85492b18fec552fc7c7e/table/iterator.go#L283)) takes place over the `baseKey` of each blocks in the SST. Once it's determined which block may contain the target key a second (lower layer) search happens inside the block where its keys are [iterated sequentially](https://github.com/dgraph-io/badger/blob/a39f9e06de643cd1edbc85492b18fec552fc7c7e/table/iterator.go#L84-L86) from start to end to find the closest key in the block to the target key. In each of those iterations the key difference stored has to be [decoded](https://github.com/dgraph-io/badger/blob/a39f9e06de643cd1edbc85492b18fec552fc7c7e/table/iterator.go#L111) back into the original key for the comparison to happen (where the decoded value is discarded in the next operation without saving the work done). This second search stage breaks the operation cost away from `O(log n)` leaning more towards `O(n)`.

# Conclusion

This analysis is by no means a criticism to the search algorithm, there is a trade-off between increasing CPU power and decreasing memory consumption which leads to better cache performance (and a memory fetch is usually much more expensive than a group simple increase-and-compare CPU instructions), but in the IPFS ecosystem this gain is not significant due to its key nature and distribution and I believe (although I don't have precise benchmarks to prove it) that the block search system in this scenario costs more than what it contributes to performance, and reducing `restartInterval` might help towards mitigating that.

I've tried to elaborate on this issue to get feedback from the core developers to check if the logic applied here (and the interpretation of the code) is correct and maybe to look for ways to mitigate this issue further (reducing `restartInterval` seemed like the simplest initial proposal but I would like to bypass the block search system as much as possible without breaking API).

**Comment by manishrjain** [comment]
Happy to make it an option. IIRC, we had picked this number from RocksDB, but I can't seem to locate it anymore.

Each SSTable being produced has these blocks, and then a block index at the end. The important thing to note here is, the closer the `restartInterval` gets to 1, the more blocks there would be, and the bigger the block index. If it reaches one, then the block index size would be bigger than the size of the rest of the data, which is obviously bad.

Have you tried benchmarking with different values of `restartInterval`? Do you see a major difference in read performance?

**Comment by schomatis** [comment]
@manishrjain Thanks for the fast reply as usual.

> Happy to make it an option.

Great, I'll prepare a PR then for you to review.

> Each SSTable being produced has these blocks, and then a block index at the end. The important thing to note here is, the closer the restartInterval gets to 1, the more blocks there would be, and the bigger the block index. If it reaches one, then the block index size would be bigger than the size of the rest of the data, which is obviously bad.

Agreed, there is definitively a trade-off between storage space and search time. Since the key-value entries in the SST are not of fixed length there needs to be a vector of positions for the binary search to happen (instead of the linear search that traverses each element one by one to find out its length).

If I understand correctly the block index (generated by the [`blockIndex`](https://github.com/dgraph-io/badger/blob/a39f9e06de643cd1edbc85492b18fec552fc7c7e/table/builder.go#L191) function) generates an extra of 4 bytes per block (roughly, without counting the extra 4 bytes for the number of blocks). Even in the extreme case of one block per key-value entry those 4 bytes might be acceptable, for IPFS with 32 byte hash keys (plus an already 10 byte entry header plus the value bytes) I would venture to say yes.

Actually that extra 4-bytes offset per entry already exists inside the header in the form of the [`prev`](https://github.com/dgraph-io/badger/blob/a39f9e06de643cd1edbc85492b18fec552fc7c7e/table/builder.go#L43) value which is only used locally inside the block (for reversed iterations) but could have been applied in a global SST scope for a general binary search across the entire SST (instead of a binary search for the blocks only and then a linear search for the key inside the block).

My main point here is that the size of the blocks might be an application-dependent decision which I think may be worth for the user to consider and adjust. (Ideally I would even prefer to have the option to forgo the block system altogether for certain use cases, something that wouldn't break the API but would certainly require more careful consideration.)

> Have you tried benchmarking with different values of restartInterval? Do you see a major difference in read performance?

I'll try to prepare some simple benchmarks to quantify the   search-key performance variation with different `restartInterval` values, I agree that if there's no performance impact there is no need to burden the user with yet another configuration option.

**Comment by manishrjain** [comment]
You can use [badger-bench](https://github.com/dgraph-io/badger-bench) for the benchmarks. It would take care of populating Badger with data, and then provide random read, iteration and other functions for benchmarking.

**Comment by schomatis** [comment]
@manishrjain Thanks for the pointer however those tests are too involved for what I need to measure right now (they aren't automated and encompass several steps to get them going). I prepared a rudimentary benchmark in another [PR](https://github.com/dgraph-io/badger/pull/447/commits/f682ed52dee11170c798dbb01ba0995507dcee88). I've tested get operations for different block sizes and the [benchmark showed](https://ci.appveyor.com/project/manishrjain/badger/build/1279#L59) that block sizes of 1/10 performed roughly 20% better than the default 100. Also, increasing the block size to 1000 severely degrades performance due to the linear search algorithm.

Compare that with a [binary search algorithm](https://github.com/dgraph-io/badger/pull/447/commits/47da3d6cb4924e0fbdfa3ac24ca420d6c4b11f49) implementation in the block iterator with an added key-value entry offset index which [stabilizes get times](https://ci.appveyor.com/project/manishrjain/badger/build/1282#L59) for the very different (orders of magnitude) block sizes.

These of course are not conclusive benchmarks, they are just put/get operations in a loop inside a test run by the AppVeyor, but the significant time differences make me inclined to believe that block size matters, specially when doing a linear search inside the block. Maybe someone over there could perform the official benchmarks to get more conclusive results.

With all this in mind:

1. Help me prepare a PR to export the block size option, the [commit](https://github.com/dgraph-io/badger/pull/447/commits/99e1b971d5961fc8b62199c3e3abfc9c568695ad) I prepared for this is suboptimal because I could not find an appropriate path to reach the global DB options from the low level builder functions. Maybe you could advise me on a more elegant solution.

2. Consider changing the block iterator seek algorithm to a binary search, there is already a [commit](https://github.com/dgraph-io/badger/pull/447/commits/47da3d6cb4924e0fbdfa3ac24ca420d6c4b11f49) towards that end that could be refined. The extra 4-bytes-per-entry space taken by the offsets can be recovered releasing (as discussed in another comment) the `prev` field from the entry headers (the reverse iteration algorithm would just operate over the index instead of jumping from `prev` to `prev`). Additionally many of the block iterator operations (e.g., `Next`) could benefit from the use of this index.

As a side note, while working on these changes I noticed other (maybe minor) performance issues worth pursuing that could be discussed in another issue, the main one is the use of the compressed keys. While I think this is an interesting feature that has a potential performance improvement in many use cases (not IPFS but consider a database with an auto-incrementing index) I think that it could be exploited some more. The keys are stored compressed (i.e., as a difference with respect to the base key) but they are compared in its full (uncompressed) form, comparing compressed (key differences) may provide some room for performance improvements, e.g., if the compacted keys were stored preloaded in a single array (in a similar way to the key index of the SST, but not exactly as there is some fragmentation in the `blockIndex` structure) there could be some potentially important cache benefits. Anyway, that is something that should be discussed (and tested) further in a different issue.

**Comment by manishrjain** [comment]
I'm looking at this change and trying to determine if it's worth working on. It's not clear -- did you see any performance improvement with binary search over the current linear search in blocks? Looking that appveyor links doesn't make it clear.

**Comment by campoy** [comment]
Hi @schomatis,

Just letting you know it seems this issue is waiting on your input.

Thanks!

**Comment by schomatis** [comment]
Haven't really followed up on this so let's close it.
